walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,100 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Iterator;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
 
+
+/*  Find time slot(s) that all the required attendees are availble
+    Include optional attendees if at least one time slot can be found
+    Otherwise return the time slots that fit just the mandatory attendees.
+*/
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Initialize a time slot (the whole day)
+    List<TimeRange> timeSlots = new ArrayList<>();
+    timeSlots.add(TimeRange.fromStartDuration(0, 24 * 60));
+
+    // Find time slots for required attendees
+    timeSlots = queryForEvent(events, timeSlots, request.getAttendees());
+    timeSlots = deleteUnqualifiedQuery(timeSlots, request.getDuration());
+
+    // Save the result in timeSlotsForRequired
+    // in case no time slot can be found when optional attendees are included
+    List<TimeRange> timeSlotsForRequired = new ArrayList<TimeRange>(timeSlots);
+
+    // Check for the availability of optional attendees
+    timeSlots = queryForEvent(events, timeSlots, request.getOptionalAttendees());
+    timeSlots = deleteUnqualifiedQuery(timeSlots, request.getDuration());
+
+    if (timeSlots.isEmpty() && !request.getAttendees().isEmpty()) {
+      return timeSlotsForRequired;
+    }
+    return timeSlots;
+  }
+
+  // Iterate through all event to check
+  // if a request attendee is attending a certain event
+  public List<TimeRange> queryForEvent(Collection<Event> events,
+      List<TimeRange> timeSlots, Collection<String> attendees) {
+    for (Iterator<Event> i = events.iterator(); i.hasNext();) {
+      Event event = i.next();
+
+      // Iterate through attendees to check the events each of them are attending
+      for (Iterator<String> j = attendees.iterator(); j.hasNext();) {
+        String attendee = j.next();
+        if (event.getAttendees().contains(attendee)) {
+
+          // Iterate through all time periods to see
+          // if any period overlaps with an event the attendees have
+          for (ListIterator<TimeRange> k = timeSlots.listIterator(); k.hasNext();) {
+            TimeRange currTime = k.next();
```

Comment:
Perhaps rename this variable for clarity? My initial expectation was that this refers to the current wall clock time.